compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeToHalfDmaCpyNd.cpp

```diff
@@ -0,0 +1,130 @@
+        llvm::any_of(dmaOp.getAsyncTokens(), [](Value token) {
+          return isa<AMDAIE::AsyncSourceTokenType>(token.getType());
+        });
+    SmallVector<Type> resultTypes = {
+        rewriter.getType<AMDAIE::AsyncTokenType>()};
+    TypeRange sourceResultTypes =
+        hasAsyncSourceToken ? TypeRange{resultTypes} : TypeRange{};
+    rewriter.setInsertionPoint(dmaOp);
+    auto sourceDma = rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+        dmaOp.getLoc(), sourceResultTypes, connectionOp, source,
+        dmaOp.getSourceMixedOffsets(), dmaOp.getSourceMixedSizes(),
+        dmaOp.getSourceMixedStrides(), dmaOp.getSourceBdId(), sourceChannelOp,
+        useNextBd, nextBd, dmaOp.getSourceBdId());
+
+    // Convert target half.
+    Value target =
+        dmaOp.getTarget() ? dmaOp.getTarget() : connectionOp.getTarget();
+    if (connectionOp.getTargetChannels().size() != 1)
+      return connectionOp.emitOpError() << "expected a single target channel";
+    auto targetChannelOp = dyn_cast<AMDAIE::ChannelOp>(
+        connectionOp.getTargetChannels()[0].getDefiningOp());
+    bool hasAsyncTargetToken =
+        llvm::any_of(dmaOp.getAsyncTokens(), [](Value token) {
+          return isa<AMDAIE::AsyncTargetTokenType>(token.getType());
+        });
+    TypeRange targetResultTypes =
+        hasAsyncTargetToken ? TypeRange{resultTypes} : TypeRange{};
+    auto targetDma = rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+        dmaOp.getLoc(), targetResultTypes, connectionOp, target,
+        dmaOp.getTargetMixedOffsets(), dmaOp.getTargetMixedSizes(),
+        dmaOp.getTargetMixedStrides(), dmaOp.getTargetBdId(), targetChannelOp,
+        useNextBd, nextBd, dmaOp.getTargetBdId());
+    if (dmaOp.getNumResults() == 1) {
+      if (sourceDma.getNumResults() == 1) {
+        rewriter.replaceUsesWithIf(
+            dmaOp.getResult(0), sourceDma.getResult(0), [&](OpOperand &use) {
+              return isa<AMDAIE::AsyncSourceTokenType>(use.get().getType()) &&
+                     isa<AMDAIE::NpuDmaWaitOp>(use.getOwner());
+            });
+      }
+      if (targetDma.getNumResults() == 1) {
+        rewriter.replaceUsesWithIf(
+            dmaOp.getResult(0), targetDma.getResult(0), [&](OpOperand &use) {
+              return isa<AMDAIE::AsyncTargetTokenType>(use.get().getType()) &&
+                     isa<AMDAIE::NpuDmaWaitOp>(use.getOwner());
+            });
+      }
+      if (!dmaOp.getResult(0).use_empty())
+        return dmaOp.emitOpError() << "should not have any uses anymore";
+    }
+    rewriter.eraseOp(dmaOp);
+    return success();
+  }
+};
+
+namespace {
+class AMDAIEControlCodeToHalfDmaCpyNdPass
+    : public impl::AMDAIEControlCodeToHalfDmaCpyNdBase<
+          AMDAIEControlCodeToHalfDmaCpyNdPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+  void runOnOperation() override;
+};
+
+void AMDAIEControlCodeToHalfDmaCpyNdPass::runOnOperation() {
```

Comment:
I think the splitting up of the `ControlcodeLowering` pass is standalone, so I would prefer a separate PR. This helps a lot with reviewing and for future reference.